Bottom: da003840aa5b76224d070c7fae566668e12dab38
Top:    6792732a4bcf1c365885aa4c88ce2e1dda5a6a1a
Author: Jason Wang <jasowang@redhat.com>
Date:   2016-12-09 10:50:33 +0800

Refresh of net-introduce-skb_direct_xmit

---

diff --git a/net/core/dev.c b/net/core/dev.c
index f4e5f0e..87a02b0 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -2902,18 +2902,18 @@ int skb_direct_xmit(struct sk_buff *skb, bool more)
 
 	HARD_TX_LOCK(dev, txq, smp_processor_id());
 	if (!netif_xmit_frozen_or_drv_stopped(txq))
-		ret = netdev_start_xmit(skb, dev, txq, more);
+		skb = dev_hard_start_xmit(skb, dev, txq, &ret);
 	HARD_TX_UNLOCK(dev, txq);
 
 	local_bh_enable();
 
 	if (!dev_xmit_complete(ret))
-		kfree_skb(skb);
+		kfree_skb_list(skb);
 
 	return ret;
 drop:
 	atomic_long_inc(&dev->tx_dropped);
-	kfree_skb(skb);
+	kfree_skb_list(skb);
 	return NET_XMIT_DROP;
 }
 EXPORT_SYMBOL(skb_direct_xmit);
